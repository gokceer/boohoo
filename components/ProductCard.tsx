import React from 'react'
import { TouchableOpacity, Text, Image, View, StyleSheet } from 'react-native'
import { useNavigation } from '@react-navigation/native'
import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import { RootStackParamList } from "../App"

import type { CommonProductObjectType } from "../App"

type ProductNavigationProp = NativeStackNavigationProp<RootStackParamList, 'ProductDetail'>;

const ProductCard = ({ id, colour, name, price, img }: CommonProductObjectType) => {
  const navigation = useNavigation<ProductNavigationProp>()

  // Display product in a card
  return (
    <TouchableOpacity
      testID='detail_btn'
      key={id}
      style={styles.card}
      onPress={() => navigation.navigate("ProductDetail", { id, colour, name, price, img })}>
      <Image testID={"product_img"} source={{ uri: img }} style={styles.image} />
      <View style={styles.infoContainer}>
        <Text style={styles.name} numberOfLines={2} ellipsizeMode='tail'>{name}</Text>
        <Text style={styles.price}>Â£ {price}</Text>
      </View>
    </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  card: {
    backgroundColor: 'white',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 8,
    marginTop: '4%',
  },
  image: {
    width: '100%',
    aspectRatio: 1,
    resizeMode: 'contain'
  },
  infoContainer: {
    padding: 16
  },
  name: {
    fontSize: 20,
    fontWeight: 'bold'
  },
  price: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 8
  }
})

export default ProductCard