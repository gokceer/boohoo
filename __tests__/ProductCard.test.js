import { Provider } from "react-redux";
import { store } from "../store";
import { fireEvent, render } from "@testing-library/react-native";
import ProductCard from "../components/ProductCard";

const mockedNavigate = jest.fn();
const mockItemProp = {
  id: 1,
  colour: "Black",
  name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
  price: 10,
  img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
};

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

const setup = (props = {}) => {
  const wrapper = render(
    <Provider store={store}>
      <ProductCard {...props} />
    </Provider>
  );
  return wrapper;
};

describe("Product Card unit tests", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it("renders correctly, snapshot test", () => {
    const tree = setup(mockItemProp);
    expect(tree).toMatchSnapshot();
  });

  it("All parts in the product card existance test", () => {
    const wrapper = setup(mockItemProp);

    expect(wrapper.getByTestId("product_img")).toBeTruthy();
    expect(
      wrapper.getByText("Black Sheet Strappy Textured Glitter Bodycon Dress")
    ).toBeTruthy();
    expect(wrapper.getByText("Â£ 10")).toBeTruthy();
  });

  it("Touchable opacity onpress navigate to ProductDetail screen test", () => {
    const wrapper = setup(mockItemProp);
    const detailButton = wrapper.getByTestId("detail_btn");

    fireEvent.press(detailButton);

    expect(mockedNavigate).toHaveBeenCalledWith("ProductDetail", mockItemProp);
  });
});
