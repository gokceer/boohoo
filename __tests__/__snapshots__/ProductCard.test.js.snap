// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card unit tests renders correctly, snapshot test 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 8,
      "justifyContent": "center",
      "marginTop": "4%",
      "opacity": 1,
    }
  }
  testID="detail_btn"
>
  <Image
    source={
      Object {
        "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
      }
    }
    style={
      Object {
        "aspectRatio": 1,
        "resizeMode": "contain",
        "width": "100%",
      }
    }
    testID="product_img"
  />
  <View
    style={
      Object {
        "padding": 16,
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      Black Sheet Strappy Textured Glitter Bodycon Dress
    </Text>
    <Text
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "600",
          "marginBottom": 8,
        }
      }
    >
      Â£ 
      10
    </Text>
  </View>
</View>
`;
